File: frontend/src/types/paper_trade.ts
export type TradeStatus =
  | 'OPEN'
  | 'TARGET_HIT'
  | 'STOP_HIT'
  | 'TREND_EXIT'
  | 'TIME_STOP'
  | 'CLOSED'

export interface ExitAlert {
  type: 'stop_hit' | 'partial_target' | 'target_hit' | 'trend_exit' | 'time_stop'
  severity: 'danger' | 'success' | 'warning' | 'info'
  message: string
}

export interface PaperTrade {
  id: string
  symbol: string
  company_name: string
  sector: string
  strategy: string
  entry_price: number
  stop_price: number
  target_price: number
  atr: number
  shares: number
  virtual_capital: number
  entry_date: string
  notes: string | null
  status: TradeStatus
  exit_price: number | null
  exit_date: string | null
  created_at: string
}

export interface PaperTradeLiveStatus {
  trade: PaperTrade
  current_price: number | null
  current_pnl: number | null
  current_pnl_pct: number | null
  r_multiple: number | null
  progress_to_target_pct: number | null
  days_open: number
  alerts: ExitAlert[]
  ema20: number | null
  ema50: number | null
}

export interface PaperTradeCreate {
  symbol: string
  company_name: string
  sector: string
  strategy: string
  entry_price: number
  stop_price: number
  target_price: number
  atr: number
  notes?: string
  virtual_capital: number
}

export interface PositionSizingResult {
  virtual_capital: number
  risk_pct: number
  risk_amount: number
  entry_price: number
  stop_price: number
  target_price: number
  stop_distance: number
  target_distance: number
  risk_reward: number
  shares: number
  capital_needed: number
  capital_pct: number
  max_loss: number
  max_gain: number
}

File: backend/app/models/paper_trade.py
from __future__ import annotations

from enum import Enum
from typing import List, Optional

from pydantic import BaseModel


class TradeStatus(str, Enum):
    OPEN = "OPEN"
    TARGET_HIT = "TARGET_HIT"
    STOP_HIT = "STOP_HIT"
    TREND_EXIT = "TREND_EXIT"
    TIME_STOP = "TIME_STOP"
    CLOSED = "CLOSED"          # manually closed by user


class ExitAlert(BaseModel):
    type: str        # "stop_hit" | "partial_target" | "target_hit" | "trend_exit" | "time_stop"
    severity: str    # "danger" | "success" | "warning" | "info"
    message: str


class PaperTrade(BaseModel):
    id: str
    symbol: str
    company_name: str
    sector: str
    strategy: str              # "pullback" | "momentum" | "vcp"
    entry_price: float
    stop_price: float
    target_price: float
    atr: float
    shares: int
    virtual_capital: float
    entry_date: str            # "YYYY-MM-DD"
    notes: Optional[str] = None
    status: TradeStatus
    exit_price: Optional[float] = None
    exit_date: Optional[str] = None
    created_at: str


class PaperTradeCreate(BaseModel):
    symbol: str
    company_name: str
    sector: str = ""
    strategy: str = "pullback"
    entry_price: float
    stop_price: float
    target_price: float
    atr: float
    notes: Optional[str] = None
    virtual_capital: float = 100000.0
    capital_deployed: Optional[float] = None


class PaperTradeLiveStatus(BaseModel):
    trade: PaperTrade
    current_price: Optional[float]
    current_pnl: Optional[float]            # ₹ profit/loss
    current_pnl_pct: Optional[float]        # % return on capital deployed
    r_multiple: Optional[float]             # pnl / initial risk amount
    progress_to_target_pct: Optional[float] # 0–100 bar fill
    days_open: int
    alerts: List[ExitAlert]
    ema20: Optional[float] = None
    ema50: Optional[float] = None


class PositionSizingResult(BaseModel):
    virtual_capital: float
    risk_pct: float = 1.0
    risk_amount: float
    entry_price: float
    stop_price: float
    target_price: float
    stop_distance: float
    target_distance: float
    risk_reward: float
    shares: int
    capital_needed: float
    capital_pct: float
    max_loss: float
    max_gain: float


class PaperTradeCloseRequest(BaseModel):
    exit_price: float
